Include/internal/pycore_code.h

```diff
@@ -24,6 +26,103 @@ struct _PyOpcache {
     char optimized;
 };
 
+
+/* PEP 659
+ * Specialization and quickening structs and helper functions
+ */
+
+typedef struct {
+    int32_t cache_count;
+    int32_t _; /* Force 8 byte size */
+} _PyEntryZero;
+
+/* Add specialized versions of entries to this union.
+ *
+ * Do not break the invariant: sizeof(SpecializedCacheEntry) == 8
+ * Preserving this invariant is necessary because:
+    - If any one form uses more space, then all must and on 64 bit machines
+      this is likely to double the memory consumption of caches
+    - The function for calculating the offset of caches assumes a 4:1
+      cache:instruction size ratio. Changing that would need careful
+      analysis to choose a new function.
+ */
+typedef union {
+    _PyEntryZero zero;
+    PyObject *object;
+} SpecializedCacheEntry;
+
+#define INSTRUCTIONS_PER_ENTRY (sizeof(SpecializedCacheEntry)/sizeof(_Py_CODEUNIT))
+
+/* Maximum size of code to quicken, in code units. */
+#define MAX_SIZE_TO_QUICKEN 5000
+
+typedef union _cache_or_instruction {
+    _Py_CODEUNIT code[1];
+    SpecializedCacheEntry entry;
+} SpecializedCacheOrInstruction;
+
+/* Get pointer to the nth cache entry, from the first instruction and n.
+ * Cache entries are index backwards, with [count-1] first in memory, and [0] last.
+ * The zeroth entry immediately precedes the instructions.
+ */
+static inline SpecializedCacheEntry *
+_GetSpecializedCacheEntry(_Py_CODEUNIT *first_instr, Py_ssize_t n)
+{
+    SpecializedCacheOrInstruction *last_cache_plus_one = (SpecializedCacheOrInstruction *)first_instr;
+    assert(&last_cache_plus_one->code[0] == first_instr);
+    return &last_cache_plus_one[-1-n].entry;
+}
+
+/* Following two functions determine the index of a cache entry from the
```

Comment:
This comment doesn't seem correct - they _take_ the index and return oparg or offset.